llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp

```diff
@@ -1828,8 +1857,17 @@ struct DSEState {
         // underlying objects is very uncommon. If it turns out to be important,
         // we can use getUnderlyingObjects here instead.
         const Value *UO = getUnderlyingObject(DefLoc->Ptr);
-        if (!isInvisibleToCallerAfterRet(UO))
-          continue;
+        if (!isInvisibleToCallerAfterRet(UO)) {
+          // EVM local begin
+          Triple TT(DefI->getFunction()->getParent()->getTargetTriple());
+          unsigned LocAS = DefLoc->Ptr->getType()->getPointerAddressSpace();
+          if (!TT.isEVM() || LocAS == 5 || LocAS == 6)
+            continue;
+
+          if (isOnPathToFunctionReturn(Def))
```

Comment:
Can we document it? Is it upstreamable?

Reply:
It reasonable to evaluate this transformation across all targets (i.e., without limiting it to EVM) to assess whether it could be upstreamed. At first glance, this appears to be EVM-specific, since in the general case it is not guaranteed that memory side effects preceding an unreachable can be safely ignored.